packages/medusa/src/services/totals.js

```diff
@@ -0,0 +1,197 @@
+import _ from "lodash"
+import { BaseService } from "medusa-interfaces"
+
+/**
+ * A service that calculates total and subtotals for orders, carts etc..
+ * @implements BaseService
+ */
+class TotalsService extends BaseService {
+  constructor({ productVariantService }) {
+    super()
+    /** @private @const {ProductVariantService} */
+    this.productVariantService_ = productVariantService
+  }
+  /**
+   * Calculates subtotal of a given cart
+   * @param {Cart} Cart - the cart to calculate subtotal for
+   * @return {int} the calculated subtotal
+   */
+  getSubTotal(cart) {
+    let subTotal = 0
+    if (!cart.items) {
+      return subTotal
+    }
+
+    cart.items.map(item => {
+      if (Array.isArray(item.content)) {
+        const temp = _.sumBy(item.content, c => c.unit_price * c.quantity)
+        subTotal += temp * item.quantity
+      } else {
+        subTotal +=
+          item.content.unit_price * item.content.quantity * item.quantity
+      }
+    })
+    return subTotal
+  }
+
+  /**
+   * Calculates either fixed or percentage discount of a variant
+   * @param {string} lineItem - id of line item
+   * @param {string} variant - id of variant in line item
+   * @param {int} variantPrice - price of the variant based on region
+   * @param {int} valye - discount value
+   * @param {string} discountType - the type of discount (fixed or percentage)
+   * @return {{ string, string, int }} triples of lineitem, variant and
+   *    applied discount
+   */
+  calculateDiscount_(lineItem, variant, variantPrice, value, discountType) {
+    if (discountType === "percentage") {
+      return {
+        lineItem,
+        variant,
+        amount: (variantPrice / 100) * value,
+      }
+    } else {
+      return {
+        lineItem,
+        variant,
+        amount: value >= variantPrice ? variantPrice : value,
+      }
+    }
+  }
+
+  /**
+   * If the discount_rule of a discount has allocation="item", then we need
+   * to calculate discount on each item in the cart. Furthermore, we need to
+   * make sure to only apply the discount on valid variants. And finally we
+   * return ether an array of percentages discounts or fixed discounts
+   * alongside the variant on which the discount was applied.
+   * @param {Discount} discount - the discount to which we do the calculation
+   * @param {Cart} cart - the cart to calculate discounts for
+   * @return {[{ string, string, int }]} array of triples of lineitem, variant
+   *    and applied discount
+   */
+  async getAllocationItemDiscounts(discount, cart) {
```

Comment:
This is a bit tricky... I think we should avoid applying discounts to bundles: i.e. 
```
if (Array.isArray(item.content)) {
   ... do nothing
}
```

Consider a situation where a store sells shirts and shorts priced 50 and 40 respectively, they also sell a bundle with 2 shirts and 1 pair of shorts for 100 instead of 140. The line item for the bundle would be:

```
title: ...
description: ...
...
content: [
   {
      variant: {shirt ...}
      product: {...}
      quantity: 2
   }
   {
      variant: {shorts ...}
      product: {...}
      quantity: 1
   }
]
quantity: 1
unit_price: 100
```

If the customer has a discount for shorts it isn't really clear which price to discount. You could get the short price and then try to discount that but it wouldn't really be clear how the new bundle price should be calculated thereafter.

Reply:
Agree.. Let's wait with this functionality.